fs-storage/src/folder_storage.rs

```diff
@@ -0,0 +1,598 @@
+                log::debug!(
+                    "Both RAM and disk modifications found, stopping check"
+                );
+                break;
+            }
+        }
+
+        // Skip this check if this divergent condition has already been reached
+        if !ram_newer || !disk_newer {
+            // Check for files on disk that aren't in RAM
+            for entry in fs::read_dir(&self.path)? {
+                let entry = entry?;
+                let path = entry.path();
+                if path.is_file()
+                    && path.extension().map_or(false, |ext| ext == "bin")
+                {
+                    let key = path
+                        .file_stem()
+                        .unwrap()
+                        .to_str()
+                        .unwrap()
+                        .parse::<K>()
+                        .map_err(|_| {
+                            ArklibError::Storage(
+                                self.label.clone(),
+                                "Failed to parse key from filename".to_owned(),
+                            )
+                        })?;
+                    if !self.ram_timestamps.contains_key(&key) {
+                        disk_newer = true;
+                        log::debug!(
+                            "Disk newer: file {} exists on disk but not in RAM",
+                            path.display()
+                        );
+                        break;
+                    }
+                }
+            }
+        }
+
+        let status = match (ram_newer, disk_newer) {
+            (false, false) => SyncStatus::InSync,
+            (true, false) => SyncStatus::StorageStale,
+            (false, true) => SyncStatus::MappingStale,
+            (true, true) => SyncStatus::Diverge,
+        };
+
+        log::info!("{} sync status is {}", self.label, status);
+        Ok(status)
+    }
+
+    /// Sync the in-memory storage with the storage on disk
+    fn sync(&mut self) -> Result<()> {
+        match self.sync_status()? {
+            SyncStatus::InSync => Ok(()),
+            SyncStatus::MappingStale => self.read_fs().map(|_| ()),
+            SyncStatus::StorageStale => self.write_fs().map(|_| ()),
+            SyncStatus::Diverge => {
+                let data = self.load_fs_data()?;
+                self.merge_from(&data)?;
+                self.write_fs()?;
+                Ok(())
+            }
+        }
+    }
+
+    /// Read the data from folder storage
+    fn read_fs(&mut self) -> Result<&BTreeMap<K, V>> {
+        let data = self.load_fs_data()?;
+        self.data = data;
+        Ok(&self.data.entries)
+    }
+
+    /// Get a value from the internal mapping
+    fn get(&self, id: &K) -> Option<&V> {
+        self.data.entries.get(id)
+    }
+
+    /// Write the data to folder
+    ///
+    /// Update the modified timestamp in file metadata to avoid OS timing issues
+    /// https://github.com/ARK-Builders/ark-rust/pull/63#issuecomment-2163882227
```

Comment:
We should not have links to discussions in the code, better to just explain the idea in simple words.